File: src/main/java/org/hunter/combinationsum.java
package org.hunter;

import java.util.*;

class combinationsum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        combinationSum(candidates, target, 0, 0, ans, new ArrayList<>());
        return ans;
    }

    void combinationSum(int [] candidates, int target, int index, int sum, List<List<Integer>> ans, List<Integer> cur){
        if(sum == target){
            ans.add(new ArrayList<>(cur));
            return;
        }
        if(sum > target){
            return;
        }
        for(int i = index; i < candidates.length; ++i){
            cur.add(candidates[i]);
            combinationSum(candidates, target, i, sum + candidates[i], ans, cur);
            cur.remove(cur.size() - 1);
        }
    }
}

File: combination_sum.py
class Solution:
    #O(n^target) where n = len(candidates), O(1)
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []

        def combo_sum(index, candidates, cur_ans, sum, target):
            if sum == target:
                ans.append(cur_ans.copy())
                return
            if sum > target:
                return
            if index > len(candidates) - 1:
                return
            cur_ans.append(candidates[index])
            combo_sum(index, candidates, cur_ans, sum + candidates[index], target)
            cur_ans.pop()
            combo_sum(index + 1, candidates, cur_ans, sum, target)

        combo_sum(0, candidates, [], 0, target)

        return ans
